From a7ecde515b30c1c2b2a4055a696429117ee7481a Mon Sep 17 00:00:00 2001
From: Zhou Song <zhous@codeaurora.org>
Date: Thu, 13 Feb 2020 15:46:35 +0800
Subject: [PATCH 36/37] Audio: use cached BT codec info during disconnecting

When disconnecting BT device, BT device state machine may be
destroyed and cause query of codec failed.
This can make device disconnection failed in low layer audio
service.

CRs-Fixed: 2604724
Change-Id: I970802310eae0b5b43340aaddb312906e1a4c444
---
 .../server/audio/AudioDeviceInventory.java        | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/audio/AudioDeviceInventory.java b/services/core/java/com/android/server/audio/AudioDeviceInventory.java
index d5388199983..b5765244a89 100644
--- a/services/core/java/com/android/server/audio/AudioDeviceInventory.java
+++ b/services/core/java/com/android/server/audio/AudioDeviceInventory.java
@@ -817,7 +817,20 @@ public class AudioDeviceInventory {
                 delay = 0;
             }
 
-            final int a2dpCodec = mDeviceBroker.getA2dpCodec(device);
+            final int a2dpCodec;
+            if (state == BluetoothA2dp.STATE_DISCONNECTED) {
+                final String key = DeviceInfo.makeDeviceListKey(AudioSystem.DEVICE_OUT_BLUETOOTH_A2DP,
+                        device.getAddress());
+                final DeviceInfo di = mConnectedDevices.get(key);
+                if (di != null) {
+                    a2dpCodec = di.mDeviceCodecFormat;
+                } else {
+                    Log.e(TAG, "invalid null DeviceInfo in setBluetoothA2dpDeviceConnectionState");
+                    return;
+                }
+            } else {
+                a2dpCodec = mDeviceBroker.getA2dpCodec(device);
+            }
 
             if (AudioService.DEBUG_DEVICES) {
                 Log.i(TAG, "setBluetoothA2dpDeviceConnectionState device: " + device
-- 
2.28.0

